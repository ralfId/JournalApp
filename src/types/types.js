export const types =  {
    
    //authReducer
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    //uiReducer
    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start Loading',
    uiFinishLoading: '[UI] Finish Loading',

    //notesReducer
    notesAddNewNote: '[Notes] Add New Note',
    notesSetActiveNote: '[Notes] Set Active Note',
    notesLoadNotes: '[Notes] Load Notes',
    notesRefrshUpdatedNote: '[Notes] Refresh Updated Note',
    notesUpdateImageNote: '[Notes] Update Image Note',
    notesDeleteNote: '[Notes] Delete Note',
    notesLogoutCleaningNotes: '[Notes] Logout Cleaning Notes',
}
