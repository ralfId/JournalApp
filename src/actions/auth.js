import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import Swal from 'sweetalert2'

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading());

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {

                dispatch(login(user.uid, user.displayName));

                dispatch(finishLoading());

            })
            .catch(e => {
                
                dispatch(finishLoading());

                Swal.fire({
                    icon: 'error',
                    title: '',
                    text: e.message,
                  })
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
        
            .then(async ({ user }) => {

                await user.updateProfile({ displayName: name })
                dispatch(login(user.uid, user.displayName))

            }).catch(error =>{
                 Swal.fire('Error', error.message, 'error')
                })
    }
}


export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => { dispatch(login(user.uid, user.displayName)) })
    }
}


export const login = (uid, displayName) => (
    {
        type: types.login,
        payload: { uid, displayName }
    })

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout())
        dispatch(notesLogout());
    }
}
export const logout = (params) => ({ type: types.logout})

