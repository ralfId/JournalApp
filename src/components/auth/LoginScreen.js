import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'


import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { removeError, setError } from '../../actions/ui'
import { useForm } from '../../hooks/useForms'

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { msgError, loading } = useSelector(state => state.ui)
    

    const [formValues, handleInputChange] = useForm({ email: 'ralf_raid@yopmail.com', password: '123456' })
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password))
        }
    }

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError('email is not valid'))
            return false
        } else if (password.length < 5) {

            dispatch(setError('password should be at least 6 characters'))
            return false;
        }

        dispatch(removeError())
        return true;

    }


    const handleGoogleLogin = () => { dispatch(startGoogleLogin()) }


    return (
        <>
            <h3 className="auth__title"> Login Screen </h3>
            <form onSubmit={handleLogin} className="animate__animated animate__fadeIn  animate__faster  ">
                {
                    msgError &&
                    (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <label htmlFor='email'>Email</label>
                <input className="auth__input" name='email' type='text' autoComplete='off' value={email} onChange={handleInputChange} />

                <label htmlFor='password'>Password</label>
                <input className="auth__input" name='password' type='password' autoComplete='off' value={password} onChange={handleInputChange} />
                <br />
                <button className="btn btn-primary btn-block" disabled={loading}>
                    SigIn
                </button>



                <div className="auth__social-networks">
                    <p>Login with social netwoks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link className="link" to="/auth/register">
                    Crate new acount
                </Link>
            </form>
        </>
    )
}
