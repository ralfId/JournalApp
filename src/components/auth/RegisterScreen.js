import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'
import { setError, removeError } from '../../actions/ui'
import { useForm } from '../../hooks/useForms'

export const RegisterScreen = () => {


    const dispatch = useDispatch();
    const {msgError} = useSelector(st => st.ui);

    const [formValues, handleInputChange] = useForm(
        {
            name: 'rafael',
            email: 'ralf_raid@yopmail.com',
            password: '123456',
            confirmpassword: '123456',
        }
    )

    const { name, email, password, confirmpassword } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0) {

            dispatch(setError('name is required'))
            return false;

        } else if (!validator.isEmail(email)) {

            dispatch(setError('email is not valid'))
            return false;

        } else if (password !== confirmpassword || password.length < 5) {

            dispatch(setError('password should be at least 6 characters an match each other'))
            return false;

        }

        dispatch(removeError())
        return true;
    }


    return (
        <>
            <h3 className="auth__title"> Register Screen </h3>
            <form onSubmit={handleRegister} className="animate__animated animate__fadeIn  animate__faster  ">
                {
                    msgError &&
                    (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <label htmlFor='name'>Name</label>
                <input className="auth__input" name='name' value={name} type='text' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='email'>Email</label>
                <input className="auth__input" name='email' value={email} type='text' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='password'>Password</label>
                <input className="auth__input" name='password' value={password} type='password' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='password'>Confirm password</label>
                <input className="auth__input" name='confirmpassword' value={confirmpassword} type='password' autoComplete='off' onChange={handleInputChange} />
                <br />
                <button className="btn btn-primary btn-block mb-5 mt-5">
                    Register
            </button>




                <Link className="link " to="/auth/login">
                    Already register?
            </Link>
            </form>
        </>
    )
}
