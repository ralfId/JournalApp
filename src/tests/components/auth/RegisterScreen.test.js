import React from 'react';
import { Provider } from 'react-redux';
import { mount } from "enzyme";
import { MemoryRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import '@testing-library/jest-dom';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }

};

let store = mockStore(initState)

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)
describe('pruebas en RegisterScreen.js', () => {

    test('debe hacer match con el snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    });

    test('debe hacer el dispatch de la accion respectiva', () => {

        //get input value
        const emailField = wrapper.find('input[name="email"]');

        //simulate onChange event
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        const actions = store.getActions()
        
        expect(actions[0]).toEqual( { type: types.uiSetError, payload: 'email is not valid' })

    });

    test('debe de mostrar la caja de texto con el error ', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'email is not valid'
            }
        
        };
        
        let store = mockStore(initState)
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
            )

           const errorBoxExist = wrapper.find('.auth__alert-error').exists();
           const errorBoxText = wrapper.find('.auth__alert-error').text().trim()

           expect(errorBoxExist).toBe(true);
           expect(errorBoxText).toBe('email is not valid');
    })
    


});