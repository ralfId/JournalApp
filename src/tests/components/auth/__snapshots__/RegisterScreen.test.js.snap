// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en RegisterScreen.js debe hacer match con el snapshot 1`] = `
Array [
  <h3
    className="auth__title"
  >
     Register Screen 
  </h3>,
  <form
    className="animate__animated animate__fadeIn  animate__faster  "
    onSubmit={[Function]}
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      type="text"
      value="rafael"
    />
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      type="text"
      value="ralf_raid@yopmail.com"
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      type="password"
      value="123456"
    />
    <label
      htmlFor="password"
    >
      Confirm password
    </label>
    <input
      autoComplete="off"
      className="auth__input"
      name="confirmpassword"
      onChange={[Function]}
      type="password"
      value="123456"
    />
    <br />
    <button
      className="btn btn-primary btn-block mb-5 mt-5"
    >
      Register
    </button>
    <a
      className="link "
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>,
]
`;
